sklearn/neighbors/_base.py

```diff
@@ -186,7 +186,7 @@ def _check_precomputed(X):
     else:
         graph = X
 
-    if graph.format not in ("csr", "csc", "coo", "lil"):
+    if graph.format not in ("csr", "csc", "coo", "lil", "bsr", "dok"):
```

Comment:
The implementation in `main` converts "bsr" to "csr" before calling `_check_precomputed`. I see two paths forward:

1. `_check_precomputed` should allow `bsr` and `dok`
2. `test_kneighbors_regressor_sparse` is incorrect in testing `bsr` and `dok`.

@TomDLT What was the issue with `bsr` and `dok` with explicit zeros?

Reply:
The neighbors module uses scipy sparse matrices using [graph semantics (glossary)](https://scikit-learn.org/stable/glossary.html#term-sparse-graph):
> As with `scipy.sparse.csgraph`, explicit zeros are interpreted as the number 0, but implicit zeros indicate a masked or absent value, such as the absence of an edge between two vertices of a graph, where an explicit value indicates an edge’s weight. This interpretation is adopted to represent connectivity in clustering, in representations of nearest neighborhoods (e.g. `neighbors.kneighbors_graph`), and for precomputed distance representation where only distances in the neighborhood of each point are required.

However, the handling of explicit zeros is inconsistent over scipy sparse formats.
- The first example shows that in the BSR format, the block structure often change implicit zeros into explicit zeros.
- The second example shows that DOK (and actually LIL as well) removes some explicit zeros, for instance zeros set through item assignment.
Therefore, BSR/DOK/LIL formats might lead to unexpected results when used as precomputed sparse graphs.

```python
import numpy as np
from scipy.sparse import (bsr_matrix, coo_matrix, csc_matrix, csr_matrix,
                          dok_matrix, lil_matrix)

functions = [
    csr_matrix, csc_matrix, coo_matrix, lil_matrix, dok_matrix, bsr_matrix
]

# BSR format is forced to store some zeros explicitly
X = np.ones((3, 3))
X[0, 0] = 0  # future implicit zero
for function in functions:
    print(function.__name__, function(X).nnz == 8)
print("")
# csr_matrix True
# csc_matrix True
# coo_matrix True
# lil_matrix True
# dok_matrix True
# bsr_matrix False


# LIL/DOK remove explicitly set zeros
for function in functions:
    X = function(np.ones((3, 3)))
    try:
        X[0, 0] = 0  # explicit zero
        print(function.__name__, X.nnz == 9)
    except Exception:  # COO/BSR do not support item assignment
        print(function.__name__, "Skip")
print("")
# csr_matrix True
# csc_matrix True
# coo_matrix Skip
# lil_matrix False
# dok_matrix False
# bsr_matrix Skip
```

As a side note, explicitly set zeros in CSR format are preserved when converting to other formats. So depending on how the sparse graph is generated, all formats might work as expected. Moreover, mistakes can be made with any format, by simply calling `X.eliminate_zeros()` to remove all explicitly stored zeros. So we could accept all formats.
```py
# All formats preserve explicit zeros when converted from CSR
for function in functions:
    X = csr_matrix(np.ones((3, 3)))
    X[0, 0] = 0  # explicit zero
    print(function.__name__, function(X).nnz == 9)
print("")
# csr_matrix True
# csc_matrix True
# coo_matrix True
# lil_matrix True
# dok_matrix True
# bsr_matrix True

```
___
However, I think raising an error highlights the danger, and users can easily cast to CSR format if they want to force their usecase. So I would suggest:
- raising for BSR
- raising for DOK and LIL, or at least be consistent between the two
- adding some tests

Reply:
(I added this error based on this suggestion https://github.com/scikit-learn/scikit-learn/pull/10482#issuecomment-359959580, but I did not double check it, and I did not implement tests concerning this decision)

Reply:
So we should rather update the previously failing tests to expect a failure for those cases. I think we can treat it like a bug / no backward compat and it won't impact many users any way.

Ideally we would add a new test based on https://github.com/scikit-learn/scikit-learn/pull/10482#issuecomment-359959580 but I would not block this PR for this.

Reply:
I updated the tests and whats new treating it as a bug fix.